src/IECoreScene/MeshAlgoNormals.cpp

```diff
@@ -42,10 +42,417 @@
+		return j * ( n - 2 ) - ( (j - 1) * j ) / 2 + i - 1;
+	}
+}
+
+PrimitiveVariable calculateNormalsImpl( const MeshPrimitive *mesh, PrimitiveVariable::Interpolation interpolation, MeshAlgo::NormalWeighting weighting, float thresholdAngle, const std::string &position, const Canceller *canceller )
+{
+	// \todo - if care about perf of this method, we could probably find a way to parallelize it
+
+	const V3fVectorData *pData = mesh->variableData<V3fVectorData>( position, PrimitiveVariable::Vertex );
+	if( !pData )
+	{
+		throw InvalidArgumentException( boost::str( boost::format( "MeshAlgo::calculateNormals : MeshPrimitive has no \"%s\" primitive variable." ) % position ) );
+	}
+	const std::vector<V3f> &points = pData->readable();
+
+	if(
+		interpolation != PrimitiveVariable::Uniform &&
+		interpolation != PrimitiveVariable::Vertex &&
+		interpolation != PrimitiveVariable::Varying &&
+		interpolation != PrimitiveVariable::FaceVarying
+	)
+	{
+		throw InvalidArgumentException( "MeshAlgo::calculateNormals : \"interpolation\" must be Vertex, Uniform or FaceVarying" );
+	}
+
+	const auto &verticesPerFace = mesh->verticesPerFace()->readable();
+
+	std::vector<int> startPerFace;
+	startPerFace.reserve( verticesPerFace.size() );
+
+	int offset = 0;
+	for( auto numVerts : verticesPerFace )
+	{
+		startPerFace.push_back( offset );
+		offset += numVerts;
+	}
+
+	const auto &vertIds = mesh->vertexIds()->readable();
+
+	if( interpolation == PrimitiveVariable::Uniform )
+	{
+		V3fVectorDataPtr faceNormalsData = new V3fVectorData;
+		faceNormalsData->setInterpretation( GeometricData::Normal );
+		auto &faceNormals = faceNormalsData->writable();
+		faceNormals.resize( verticesPerFace.size() );
+
+		tbb::parallel_for( tbb::blocked_range<int>( 0, verticesPerFace.size() ),
+			[ &verticesPerFace, &startPerFace, &vertIds, &points, &faceNormals, &canceller ]
+			( const tbb::blocked_range<int> &range )
+			{
+				for( int faceId = range.begin(); faceId != range.end(); ++faceId )
+				{
+					Canceller::check( canceller );
+
+					int numVerts = verticesPerFace[faceId];
+					int faceStart = startPerFace[faceId];
+
+					faceNormals[faceId] = polygonNormal(
+						PolygonVertexIterator( &vertIds[faceStart], &points[0] ),
+						PolygonVertexIterator( &vertIds[faceStart + numVerts], &points[0] )
+					);
+				}
+			}
+		);
+		return PrimitiveVariable( interpolation, faceNormalsData );
+	}
+
+	V3fVectorDataPtr faceVertexNormalsData = new V3fVectorData;
+	faceVertexNormalsData->setInterpretation( GeometricData::Normal );
+	auto &faceVertexNormals = faceVertexNormalsData->writable();
+	faceVertexNormals.resize( vertIds.size() );
+
+	std::vector< float > normalWeights;
+	if( interpolation == PrimitiveVariable::FaceVarying )
+	{
+		// When using FaceVarying interpolation, we need access to the normalized normals, so we can't
+		// bake in the weighting
+		normalWeights.resize( vertIds.size() );
+	}
+
+	tbb::parallel_for( tbb::blocked_range<int>( 0, verticesPerFace.size() ),
```

Comment:
Might be easiest to do the `this_task_arena_isolate` part once on the outside, where we call into `calculateNormalsImpl()`, rather than add another level of nesting inside the function.